src/unix/tty.c

```diff
@@ -335,6 +335,37 @@ int uv_tty_set_mode(uv_tty_t* tty, uv_tty_mode_t mode) {
 }
 
 
+void uv__tty_close(uv_tty_t* handle) {
+  int expected;
+  int fd;
+
+  fd = handle->io_watcher.fd;
+  if (fd == -1)
+    goto done;
+
+  /* This is used for uv_tty_reset_mode() */
+  do
+    expected = 0;
+  while (!atomic_compare_exchange_strong(&termios_spinlock, &expected, 1));
+
+  if (fd == orig_termios_fd) {
+    /* XXX(bnoordhuis) the tcsetattr is probably wrong when there are still
+     * other uv_tty_t handles active that refer to the same tty/pty but it's
+     * hard to recognize that particular situation without maintaining some
+     * kind of process-global data structure, and that still won't work in a
+     * multi-process setup.
+     */
+    uv__tcsetattr(fd, TCSANOW, &orig_termios);
```

Comment:
Add a conditional here to fix the XXX comment above?
```suggestion
    if (tty->mode != UV_TTY_MODE_NORMAL)
        uv__tcsetattr(fd, TCSANOW, &orig_termios);
```

Reply:
That won't fix the issue. In an ideal world we only reset the tty when we're the final "owner" (so to speak) but we can't really know when that is.

Reply:
True, not a fix in the absolute sense (we would have to store the ptyN info that libuv computed earlier for that and ref count those to make this fully correct), but it would hopefully minimize surprises?

Reply:
Now that I'm hale and healthy again... checking `tty->mode` is a race condition because it's set elsewhere _after_ the critical section, with a tcsetattr call in between to boot.

If `fd == orig_termios_fd`, then it should always be okay to apply `orig_termios`. Worst case it's a no-op when a signal handler calls `uv_tty_reset_mode()` when execution is inside `uv_tty_set_mode()`, after the critical section but before the mode field is updated.

(We could roll everything in `uv_tty_set_mode()` into a big critical section but that's unnecessary, just the `fd == orig_termios_fd` condition here is enough to ensure correctness.)